lualib/lua_redis.lua

```diff
@@ -146,6 +156,89 @@ end
 exports.rspamd_parse_redis_server = rspamd_parse_redis_server
 exports.parse_redis_server = rspamd_parse_redis_server
 
+local function get_key_indexes(cmd, args)
+  local idx_l = {}
+  local process_cmd = {
+    mget = function()
+      for i = 1, #args do
+        table.insert(idx_l, i)
+      end
+    end,
+    set = function()
+      idx_l = {1}
+    end,
+    eval = function()
+      local numkeys = args[2]
+      if numkeys >= 1 then
+        for i = 3, numkeys + 2 do
+          table.insert(idx_l, i)
+        end
+      end
+    end,
+    script = function() end
+  }
+  process_cmd.get = process_cmd.set
+  process_cmd.hset = process_cmd.set
+  process_cmd.hget = process_cmd.set
+  process_cmd.hgetall = process_cmd.set
+  process_cmd.hdel = process_cmd.set
+  process_cmd.hincrby = process_cmd.set
+  process_cmd.hincrbyfloat = process_cmd.set
+  process_cmd.hscan = process_cmd.set
+  process_cmd.evalsha = process_cmd.eval
+  cmd = string.lower(cmd)
+  if process_cmd[cmd] then
+    process_cmd[cmd]()
+  else
+    logger.warnx(rspamd_config, "Don't know how to extract keys for %s Redis command", cmd)
+  end
+  return idx_l
+end
+
+local function get_key_expansion_metadata(task)
+  local lazy_meta = {}
+  local gen_meta = {
+    principal_recipient = function()
+      local a = (task:get_principal_recipient() or E)['addr']
+      if a and string.len(a) == 0 then
+        a = '<>'
+      end
+      return a
+    end,
+    principal_recipient_domain = function()
+      return (task:get_principal_recipient() or E)['domain']
+    end,
+    ip = function()
+      local i = task:get_ip()
+      if i and i:is_valid() then return i:to_string() end
+    end
```

Comment:
Will want to add a couple more functions here I suppose: `smtp_from`, `asn`, whatever else is likely to be useful (unsure).